### PR TITLE
refactor/legend-states

### PR DESCRIPTION
**Pull request by ff6347**
This should eventually replace

- rainVisible
- pumpsVisible
- treesVisible

### CODE REVIEW

File: src/common/interfaces.ts [hunk 1]
```diff
@@ -97,3 +95,25 @@ export interface Tree {
...
   caretaker?: string | null;
 }
+
+// Legend shared
+export interface StyledProps {
+  active?: boolean;
+}
+
+export interface FlexColumnProps {
+  isLast?: boolean;
```
Comment (vogelino) [inline src/common/interfaces.ts:105]: Same for all of the inserted interfaces in this file. This file is not a place to simply transfer all the types into. It should only contain the minimum and only business logic relate `interfaces` like `TreeType` or `UserType`.

File: src/common/interfaces.ts [hunk 2]
```diff
@@ -54,9 +54,7 @@ export interface StoreProps {
   wateredTrees: string[];
   adoptedTrees: Pick<Tree, 'id'>[];
 
-  treesVisible: boolean;
-  pumpsVisible: boolean;
-  rainVisible: boolean;
+  visibleLayer: 'rain' | 'trees' | 'pumps';
```
Comment (vogelino) [inline src/common/interfaces.ts:57]: How about making the name even more explicit: `visibleMapLayer` ?

File: src/components/Legend/Flex.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import styled from 'styled-components';
+
+export const Flex = styled.div`
```
Comment (vogelino) [inline src/components/Legend/Flex.tsx:3]: What is a `Flex`? I think the name is not really explanatory, even when thinking about flexbox. When do I use this Component and why does it set some borders?

File: src/components/Legend/FlexRow.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import styled from 'styled-components';
+
+export const FlexRow = styled.div`
```
Comment (vogelino) [inline src/components/Legend/FlexRow.tsx:3]: Same here. These components have very generic names but very precise stylings. Please rather put all of your StyledComponents above the components that use it. Unless they are used by many components in which case it should be more explicit what it is made for.

File: src/components/Legend/Legend.tsx [hunk 5]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
-            <PumpsDot color={lockedColor.hex} size={13} />
-            <PumpLabel>{'verriegelt'}</PumpLabel>
-          </FlexRow>
-
-          <FlexRow>
-            <PumpsDot color={defaultColor.hex} size={13} />
-            <PumpLabel>{'unbekannt'}</PumpLabel>
-          </FlexRow>
+          {visibleLayer === 'trees' && <TreesColorLegend />}
+          {visibleLayer === 'rain' && <RainColorLegend />}
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:159]: This is much better!

File: src/components/Legend/Legend.tsx [hunk 6]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+
+      <FlexColumn isLast={true}>
+        <StyledCardDescription>Datenpunkte</StyledCardDescription>
+        <StyledCardDescriptionSecond>
+          durch Klick ein- & ausblenden.
+        </StyledCardDescriptionSecond>
+        <UnstyledFlexWidth
+          active={visibleLayer === 'trees' ? true : false}
+          onClick={() => {
+            store.setState({ visibleLayer: 'trees' });
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:172]: There should be an action to set the visible layer. Please don't use `store.setState`.

File: src/components/map/index.tsx [hunk 7]
```diff
@@ -142,7 +134,7 @@ class DeckGLMap extends React.Component<DeckGLPropType, DeckGLStateType> {
           }
         },
         getLineColor: [247, 105, 6, 255],
-        visible: treesVisible,
+        visible: visibleLayer === 'trees' ? true : false,
```
Comment (vogelino) [inline src/components/map/index.tsx:137]: The condition `visibleLayer === 'trees'` already returns a boolean. No need to use a ternary operator here.

File: src/components/map/index.tsx [hunk 8]
```diff
@@ -436,7 +428,8 @@ class DeckGLMap extends React.Component<DeckGLPropType, DeckGLStateType> {
 
   _updateStyles(prevProps: DeckGLPropType): void {
     if (map && isMobile) {
-      if (!this.props.treesVisible) {
+      // if (!this.props.treesVisible) {
```
Comment (vogelino) [inline src/components/map/index.tsx:431]: Please remove this comment.

File: src/state/Store.tsx [hunk 9]
```diff
@@ -1,18 +1,16 @@
 import createStore from 'unistore';
 import { StoreProps } from '../common/interfaces';
 
-const initialState = {
+const initialState: StoreProps = {
```
Comment (vogelino) [inline src/state/Store.tsx:4]: Good one!

File: webpack/dev.config.js [hunk 10]
```diff
@@ -34,4 +34,9 @@ module.exports = merge(common, {
     },
   },
   plugins: [new ErrorOverlayPlugin(), new webpack.ProgressPlugin()],
+  resolve: {
+    alias: {
+      'react-dom': '@hot-loader/react-dom',
```
Comment (vogelino) [inline webpack/dev.config.js:39]: So niiiiiice!

File: src/components/Legend/PumpsColorLegend.tsx [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import { PumpLabel } from './PumpLabel';
+import { UnstyledFlex } from './UnstyledFlex';
+import { FlexRow } from './FlexRow';
+
+export const PumpsColorLegend: React.FC = () => {
+  return (
+    <UnstyledFlex>
+      <FlexRow>
+        <PumpsDot color={workingColor.hex} size={13} />
+        <PumpLabel>{'funktionsfähig'}</PumpLabel>
```
Comment (vogelino) [inline src/components/Legend/PumpsColorLegend.tsx:18]: No need to wrap the string into curly braces and apostrophes. Write `<PumpLabel>funktionsfähig</PumpLabel>` instead

File: src/components/Legend/Legend.tsx [hunk 12]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+          <StyledItemLabel>Öffentl. Pumpen</StyledItemLabel>
+        </UnstyledFlexWidth>
+        <UnstyledFlexWidth
+          active={visibleLayer === 'rain' ? true : false}
+          onClick={() => {
+            store.setState({ visibleLayer: 'rain' });
+          }}
+        >
+          {visibleLayer === 'rain' ? (
+            <LegendRect gradient={rainGradient} />
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:206]: This can be inlined

File: src/components/Legend/Legend.tsx [hunk 13]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+        >
+          {visibleLayer === 'pumps' ? (
+            <StrokedLegendDot gradient={`${workingColor.hex}`} />
+          ) : (
+            <StrokedLegendDot />
+          )}
+          <StyledItemLabel>Öffentl. Pumpen</StyledItemLabel>
+        </UnstyledFlexWidth>
+        <UnstyledFlexWidth
+          active={visibleLayer === 'rain' ? true : false}
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:200]: Unnecessary ternary

File: src/components/Legend/Legend.tsx [hunk 14]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+          <StyledItemLabel>Straßen- & Anlagenbäume</StyledItemLabel>
+        </UnstyledFlexWidth>
+        <UnstyledFlexWidth
+          active={visibleLayer === 'pumps' ? true : false}
+          onClick={() => {
+            store.setState({ visibleLayer: 'pumps' });
+          }}
+        >
+          {visibleLayer === 'pumps' ? (
+            <StrokedLegendDot gradient={`${workingColor.hex}`} />
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:193]: This can be inlined

File: src/components/Legend/Legend.tsx [hunk 15]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+        <StyledCardDescriptionSecond>
+          durch Klick ein- & ausblenden.
+        </StyledCardDescriptionSecond>
+        <UnstyledFlexWidth
+          active={visibleLayer === 'trees' ? true : false}
+          onClick={() => {
+            store.setState({ visibleLayer: 'trees' });
+          }}
+        >
+          {visibleLayer === 'trees' ? (
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:175]: This can be inlined

File: src/components/Legend/Legend.tsx [hunk 16]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+              color={'#2c303b'}
+              gradient={rainGradient}
+            />
+          ) : (
+            <LegendDot className={'legend-dot'} color={'#2c303b'} />
+          )}
+          <StyledItemLabel>Straßen- & Anlagenbäume</StyledItemLabel>
+        </UnstyledFlexWidth>
+        <UnstyledFlexWidth
+          active={visibleLayer === 'pumps' ? true : false}
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:187]: Unnecessary ternary

File: src/components/Legend/Legend.tsx [hunk 17]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+          {visibleLayer !== 'pumps' && (
             <StyledCardDescriptionSecond>
               der letzten 30 Tage (Liter)
             </StyledCardDescriptionSecond>
           )}
         </FlexColumn>
-        <StyledToggle onClick={() => setLegendExpanded(!legendExpanded)}>
-          {legendExpanded ? '—' : '+'}
+        <StyledToggle onClick={() => setLegendExpanded(false)}>
+          {'—'}
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:153]: Unnecessary wrapping into curly braces and quotes

File: src/components/Legend/Legend.tsx [hunk 18]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+          </StyledToggle>
+        </FlexSpace>
+      </LegendDiv>
+    );
+  }
+
   return (
-    <LegendDiv active={legendExpanded}>
-      <FlexSpace active={legendExpanded}>
+    <LegendDiv active>
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:140]: Prefer `isActive` to `active` as it implies the use of a boolean more explicitly and makes the code more understandable.

File: src/components/Legend/Legend.tsx [hunk 19]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+    return (
+      <LegendDiv>
+        <FlexSpace>
+          <FlexColumn>
+            <StyledCardDescription onClick={() => setLegendExpanded(true)}>
+              {'Legende'}
+            </StyledCardDescription>
+          </FlexColumn>
+          <StyledToggle onClick={() => setLegendExpanded(true)}>
+            {'+'}
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:132]: Unnecessary wrapping into curly braces and quotes

File: src/components/Legend/Legend.tsx [hunk 20]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
+
   const [legendExpanded, setLegendExpanded] = useState<boolean>(true);
 
+  if (legendExpanded === false) {
+    return (
+      <LegendDiv>
+        <FlexSpace>
+          <FlexColumn>
+            <StyledCardDescription onClick={() => setLegendExpanded(true)}>
+              {'Legende'}
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:128]: Unnecessary wrapping into curly braces and quotes

File: src/components/Legend/Legend.tsx [hunk 21]
```diff
@@ -179,203 +103,113 @@ const StyledCardDescriptionSecond = styled(CardDescription)`
...
-    i++;
-  }
-  gradient += ')';
-  return gradient;
-}
-const rainColors = legendArray.map(item => interpolateColor(item.value));
+const rainColors = legendLabels.map(item => interpolateColor(item.value));
 const rainGradient = createCSSGradient(rainColors);
 
 const Legend: FC = () => {
```
Comment (vogelino) [inline src/components/Legend/Legend.tsx:117]: What about renaming that to `MapLayersLegend` ?

**Review by vogelino — CHANGES_REQUESTED** [review]
Great improvement! I added small comments here and there. Most of them are cosmetics and can be ignored if you want. Regarding the `StyledComponents` I would try to refactor them so that
1. their names are more understandable
2. they are not in their own files but rather where they are used. Even if it makes a file longer it makes it more explicit and easy to refactor.
Also, please avoid unnecessary composition of styles (eg. `styled(OtherThing)`) and inter-dependencies of files (eg. having files with only `StyledComponents` inside them).

**Comment by ff6347** [comment]
@vogelino I updated the PR based on your feedback. I still moved some styled components to their own (group) files, since they are used in the different Legend states. 

I also tried to make the names a little more meaningful but to get the "right right right" name it would mean to inspect also the flex props on each one. I guess now it is better then before. :-/

**Comment by vogelino** [comment]
> I updated the PR based on your feedback...

@ff6347 Ok great thx I'll check it out

File: src/common/interfaces.ts [hunk 22]
```diff
@@ -54,9 +54,7 @@ export interface StoreProps {
   wateredTrees: string[];
   adoptedTrees: Pick<Tree, 'id'>[];
 
-  treesVisible: boolean;
-  pumpsVisible: boolean;
-  rainVisible: boolean;
+  visibleMapLayer: 'rain' | 'trees' | 'pumps';
```
Comment (vogelino) [inline src/common/interfaces.ts:57]: 👍

File: src/common/interfaces.ts [hunk 23]
```diff
@@ -97,3 +95,5 @@ export interface Tree {
   kennzeich?: string | null;
   caretaker?: string | null;
 }
+
+// Legend shared
```
Comment (vogelino) [inline src/common/interfaces.ts:99]: Forgotten comment left here

File: src/components/Legend/LegendFlexBoxes.tsx [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+import { IsActiveProps } from './MapLayersLegend';
+export interface FlexColumnProps {
+  isLast?: boolean;
+}
+export const FlexColumnLast = styled.div<FlexColumnProps>`
+  display: flex;
+  flex-direction: column;
+  ${props => (props.isLast === true ? 'margin-top:auto;' : '')}
+`;
+export const FlexRow = styled.div`
```
Comment (vogelino) [inline src/components/Legend/LegendFlexBoxes.tsx:11]: Mah...Still not convinced about their naming here but maybe that's something for another time.

**Review by vogelino — APPROVED** [review]
Great! Much better! Thank you 🙏

File: src/components/Legend/LegendFlexBoxes.tsx (hunk 24 above)
Comment (ff6347) [reply src/components/Legend/LegendFlexBoxes.tsx:11]: Me neither, but as I said. I'm actually not sure what would be the right naming. These are just some layout classes nearly no logic here.

**Comment by ff6347** [comment]
But I broke the legend 😿

**Comment by ff6347** [comment]
Fixed the legend. I'll go ahead and merge this one.

**Pull request merged**